instrumentation/servlet/servlet-common/library/src/main/java/io/opentelemetry/instrumentation/servlet/BaseServletHelper.java

```diff
@@ -0,0 +1,128 @@
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.servlet.AppServerBridge;
+import io.opentelemetry.instrumentation.api.servlet.MappingResolver;
+import io.opentelemetry.instrumentation.api.servlet.ServerSpanNaming;
+import io.opentelemetry.instrumentation.api.servlet.ServletContextPath;
+import io.opentelemetry.instrumentation.api.tracer.HttpServerTracer;
+import io.opentelemetry.instrumentation.servlet.naming.ServletSpanNameProvider;
+
+public abstract class BaseServletHelper<REQUEST, RESPONSE> {
+  protected final Instrumenter<ServletRequestContext<REQUEST>, ServletResponseContext<RESPONSE>>
+      instrumenter;
+  protected final ServletAccessor<REQUEST, RESPONSE> accessor;
+  private final ServletSpanNameProvider<REQUEST> spanNameProvider;
+
+  protected BaseServletHelper(
+      Instrumenter<ServletRequestContext<REQUEST>, ServletResponseContext<RESPONSE>> instrumenter,
+      ServletAccessor<REQUEST, RESPONSE> accessor) {
+    this.instrumenter = instrumenter;
+    this.accessor = accessor;
+    this.spanNameProvider = new ServletSpanNameProvider<>(accessor);
+  }
+
+  protected Context startSpan(
+      Context parentContext,
+      ServletRequestContext<REQUEST> requestContext,
+      ServerSpanNaming.Source namingSource) {
+    Context context = instrumenter.start(parentContext, requestContext);
+
+    REQUEST request = requestContext.request();
+    SpanContext spanContext = Span.fromContext(context).getSpanContext();
+    // we do this e.g. so that servlet containers can use these values in their access logs
+    // TODO: These are only available when using servlet instrumentation or when server
+    //  instrumentation extends servlet instrumentation e.g. jetty. Either remove or make sure they
+    //  also work on tomcat and wildfly.
+    accessor.setRequestAttribute(request, "trace_id", spanContext.getTraceId());
+    accessor.setRequestAttribute(request, "span_id", spanContext.getSpanId());
+
+    context = ServerSpanNaming.init(context, namingSource);
+    context = addServletContextPath(context, request);
+    context = customizeContext(context, request);
+
+    attachServerContext(context, request);
+
+    return context;
+  }
+
+  /** Override in subclass to customize context that is returned by {@code startSpan}. */
+  protected Context customizeContext(Context context, REQUEST request) {
+    return context;
+  }
+
+  private Context addServletContextPath(Context context, REQUEST request) {
+    String contextPath = accessor.getRequestContextPath(request);
+    if (contextPath != null && !contextPath.isEmpty() && !contextPath.equals("/")) {
+      return context.with(ServletContextPath.CONTEXT_KEY, contextPath);
+    }
+    return context;
```

Comment:
Idea for another PR: this could be encapsulated in `ServletContextPath`, kind of like `ServerSpanNaming` does (we should not expose `ContextKey`s as public vars)